mlx/primitives.cpp

```diff
@@ -2849,6 +2849,86 @@ bool Slice::is_equivalent(const Primitive& other) const {
       end_indices_ == s_other.end_indices_ && strides_ == s_other.strides_);
 }
 
+std::pair<std::vector<array>, std::vector<int>> SliceUpdate::vmap(
```

Comment:
This is great but it assumes that only `inputs[0]` is vmapped. For instance the following pseudocode should work but it won't in this case

```
def foo(x, y):
    x[:5] = y
    return x

x = mx.zeros((2, 10))
y = mx.ones((2, 5))
x = mx.vmap(foo)(x, y)
print(x)
array([[1, 1, 1, 1, 1, 0, 0, 0, 0, 0],
       [1, 1, 1, 1, 1, 0, 0, 0, 0, 0]])
```

It is also not very clear to me what is the correct behavior here when only 1 of the inputs is vmapped. Your implementation is assuming that the update was defined for one input and broadcasts it appropriately. I think this is not correct, I would just do nothing to the other input if it was not broadcasted.

Also if both are vmapped then we need to make sure that the vectorized dimensions are in the correct place so I would transpose the update vectorized axis to `ax` if they are different.

wdyt?